Source/CarthageKitTests/Fakes/FakeURLSession.swift

```diff
@@ -0,0 +1,41 @@
+import Foundation
+
+final class FakeDownloadTask: NSURLSessionDownloadTask {
+	var started = false
+	override func resume() {
+		self.started = true
+	}
+
+	var cancelled = false
+	override func cancel() {
+		self.cancelled = true
+	}
+}
+
+final class FakeDataTask: NSURLSessionDataTask {
+	var started = false
+	override func resume() {
+		self.started = true
+	}
+
+	var cancelled = false
+	override func cancel() {
+		self.cancelled = true
+	}
+}
+
+final class FakeURLSession: NSURLSession {
```

Comment:
I try to avoid DI when possible because I find it can make it harder to understand or change code. I gave [a talk](https://github.com/mdiep/composition-talk/blob/master/The%20Composition%20and%20Decomposition%20of%20Software.md) that touched on this last year. But the gist is that I try to structure programs like this:

 1. A functional core of pure functions that can easily be unit tested with simple inputs and outputs
 2. A DI-compatible layer of functions that chain things together
 3. An outer layer that connects the different pieces

I see layer 2 as being optional, depending on the project in question. Integration tests serve to test layer 3, but can be much reduced in scope if the first two layers are well tested. (I think you and I are in agreement on that point at least.)

[This article](http://jerrydantonio.com/testing-concurrent-code/) about testing concurrent code in Ruby is similar in approach, but probably more directly applicable to this particular example.

The code that we're trying to test is here:

```swift
private func fetchAllPages(URL: NSURL, _ authorizationHeaderValue: String?) -> SignalProducer<NSData, CarthageError> {
	let request = createGitHubRequest(URL, authorizationHeaderValue)

	return NSURLSession.sharedSession().rac_dataWithRequest(request)
		.mapError(CarthageError.NetworkError)
		.flatMap(.Concat) { data, response -> SignalProducer<NSData, CarthageError> in
			let thisData: SignalProducer<NSData, CarthageError> = SignalProducer(value: data)

			if let HTTPResponse = response as? NSHTTPURLResponse {
				let statusCode = HTTPResponse.statusCode
				if statusCode > 400 && statusCode < 600 && statusCode != 404 {
					return thisData
						.attemptMap { data -> Result<AnyObject, CarthageError> in
							if let object = try? NSJSONSerialization.JSONObjectWithData(data, options: []) {
								return .Success(object)
							} else {
								return .Failure(.ParseError(description: "Invalid JSON in API error response \(data)"))
							}
						}
						.map { (dictionary: AnyObject) -> String in
							if let error: GitHubError = decode(dictionary) {
								return error.message
							} else {
								return (NSString(data: data, encoding: NSUTF8StringEncoding) ?? NSHTTPURLResponse.localizedStringForStatusCode(statusCode)) as String
							}
						}
						.flatMapError { (error: CarthageError) in
							return SignalProducer(value: error.description)
						}
						.attemptMap { message -> Result<NSData, CarthageError> in
							return Result.Failure(CarthageError.GitHubAPIRequestFailed(message))
						}
				}
				
				if let linkHeader = HTTPResponse.allHeaderFields["Link"] as? String {
					let links = parseLinkHeader(linkHeader)
					for (URL, parameters) in links {
						if parameters.contains("rel=\"next\"") {
							// Automatically fetch the next page too.
							return thisData.concat(fetchAllPages(URL, authorizationHeaderValue))
						}
					}
				}
			}

			return thisData
		}
}
```

It does a few things:

1. Creates a request
2. Makes the request
3. Interprets the result

We don't need to test 2—`rac_dataWithRequest` is provided by a framework and should be tested there. We only need to test 1 (`createGitHubRequest`) and 3, which is currently embedded in the function. If we extract 3 to a separate function, we could end up with something like this:

```swift
private func fetchAllPages<O: ObjectType>(URL: NSURL, _ authorizationHeaderValue: String?) -> SignalProducer<O, CarthageError> {
    let request = createGitHubRequest(URL, authorizationHeaderValue)
    return NSURLSession.sharedSession().rac_dataWithRequest(request)
        .mapError(CarthageError.NetworkError)
        .map(O.init)
        .flatMap(.Concat) { object in
            if (object.nextPage) {
                return SignalProducer(value: object).concat(fetchAllPages(object.nextPage, authorizationHeader))
            }
            return SignalProducer(value: object)
        }
}
```

Now we can unit test `createGitHubRequest`, `O.init` for our object types, `O.nextPage`, and `fetchAllPages` doesn't need much testing. We can probably survive without testing it because it'll be obvious if it's completely broken, or we can add one or two integration-style tests (I would use an HTTP stubbing library for this).

What do you think of that approach?

> I have plenty of things I want to do for Carthage, and more coming along the way. I have an entire desk here at work that is just cards representing things I want to do in Carthage (by the way, how would you like me to communicate this? Individual pull requests for things as I do them might get annoying after a while. I can create an issue for everything, but then the graph of issues (these two refactors depend upon this refactor, this other refactor depends upon two separate refactors, etc.) will get lost. I could also create a single giant issue, and have each individual refactor as a item in a list on that issue. If you have another suggestion, then that's great too. I'd like to be able to get feedback on some proposed refactors before I actually do them.

That's awesome! :zap::zap::zap:

I'm going to add you as a contributor to Carthage. I think opening an issue for everything upfront will be overwhelming. I think a single giant issue would work (using a task list), or you could open up separate issues but not open the issue until you're nearly ready to work on it. That way you can get feedback, but you don't open 100 issues this week. :smile: